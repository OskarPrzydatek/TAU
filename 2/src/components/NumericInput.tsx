import React from 'react';

type NumericInputProps = {
  label: string;
  dataTestID?: string;
};

const NumericInput = React.forwardRef<HTMLInputElement, NumericInputProps>(
  ({ label, dataTestID }, ref) => {
    const preventSigns = ['.', '+', 'i', 'I', 'e', 'E'];

    const handleEventPreventBySign: React.KeyboardEventHandler<
      HTMLInputElement
    > = (event) => preventSigns.includes(event.key) && event.preventDefault();

    return (
      <label>
        <span>{label}</span>
        <input
          data-testid={dataTestID}
          ref={ref}
          type="number"
          onKeyDown={handleEventPreventBySign}
        />
      </label>
    );
  }
);

export default NumericInput;
