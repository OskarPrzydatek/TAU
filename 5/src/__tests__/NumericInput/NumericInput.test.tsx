import { createEvent, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import NumericInput from '../../components/NumericInput';

describe('NumericInput', () => {
  test('Component Snapshot', () => {
    const tree = renderer.create(<NumericInput label="A: " />);
    expect(tree).toMatchSnapshot();
  });

  test('ensure numeric input takes only numbers', () => {
    render(<NumericInput label="A: " dataTestID="numeric-input" />);
    const numericInput: HTMLInputElement = screen.getByTestId('numeric-input');
    fireEvent.change(numericInput, { target: { value: '1' } });
    expect(numericInput.value).toBe('1');
    fireEvent.change(numericInput, { target: { value: '.' } });
    expect(numericInput.value).toBe('');
    fireEvent.change(numericInput, { target: { value: 'E' } });
    expect(numericInput.value).toBe('');
  });

  test('ensure event prevent works correctly', () => {
    render(<NumericInput label="A: " dataTestID="numeric-input" />);
    const numericInput: HTMLInputElement = screen.getByTestId('numeric-input');
    const eventPreventSignKey = createEvent.keyDown(numericInput, { key: '.' });
    const eventNumericKey = createEvent.keyDown(numericInput, { key: '3' });
    eventPreventSignKey.preventDefault = jest.fn();
    eventNumericKey.preventDefault = jest.fn();
    fireEvent(numericInput, eventPreventSignKey);
    expect(eventPreventSignKey.preventDefault).toHaveBeenCalled();
    fireEvent(numericInput, eventNumericKey);
    expect(eventNumericKey.preventDefault).not.toHaveBeenCalled();
  });

  test('ensure handling input value via ref works correctly', () => {
    const inputRef = React.createRef<HTMLInputElement>();
    render(
      <NumericInput ref={inputRef} label="A: " dataTestID="numeric-input" />
    );
    fireEvent.change(screen.getByTestId('numeric-input'), {
      target: { value: '123' },
    });
    expect(inputRef.current?.value).toBe('123');
  });
});

export {};
