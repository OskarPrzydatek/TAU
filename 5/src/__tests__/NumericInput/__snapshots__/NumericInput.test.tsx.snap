// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput Component Snapshot 1`] = `
<label>
  <span>
    A: 
  </span>
  <input
    onKeyDown={[Function]}
    type="number"
  />
</label>
`;
